### PR TITLE
Add area (zone) to esphome core config to be suggested through API and MQTT.

### PR DESCRIPTION
**Pull request by Cossid**
# What does this implement/fix?

Add area (zone) to esphome core config to be suggested through API and MQTT.

Related PRs:
- https://github.com/esphome/aioesphomeapi/pull/607
- https://github.com/home-assistant/core/pull/102834

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3297

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040
- [X] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
esphome:
  name: light-switch
  area: Living Room

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Cossid** [comment]
Note, this add a `suggested_area` for MQTT, and adds a `area` value to the API message/DeviceInfo, but I do not currently have a dev environment for Home Assistant, so this will need changes on the ESPHome API integration in Home Assistant to consume the new variable.

File: esphome/components/api/api_pb2.h
Comment (jesserockz) [inline esphome/components/api/api_pb2.h:1]: As per the comment in the header of this file, it is autogenerated from `api.proto` with the mentioned script

**Review by jesserockz — CHANGES_REQUESTED** [review]
Onboard with the change, but the api side needs to be done properly with api.proto.

**Comment by Cossid** [comment]
Looking into it.

**Comment by Cossid** [comment]
@jesserockz I can't seem to get protobuf-compiler to install, even in a virtual environment via pip.  I have adjusted the proto file and manually adjusted the file it would have compiled accordingly.  Is there a better way to process/verify this?

**Comment by jesserockz** [comment]
The devcontainer/ `ghcr.io/esphome/esphome:lint` should have the tooling installed

**Comment by Cossid** [comment]
`:lint` does not appear to be valid, and `:dev` does not appear to have protobuf-compiler

**Comment by jesserockz** [comment]
Sorry, I typed it from memory while out, its `ghcr.io/esphome/esphome-lint:dev`

**Comment by Cossid** [comment]
Thanks for updating, I wasn't finding `protoc` available in esphome-lint:dev either.
Curious, why 16 vs 15?

**Comment by jesserockz** [comment]
15 is taken already by the `bluetooth_proxy_feature_flags`

**Comment by Cossid** [comment]
Ah, it was out of order, I see, thanks!

**Comment by jesserockz** [comment]
Just to match mqtt, ive changed the protoc name

**Comment by Cossid** [comment]
Sure, I was on the fence about that myself.

**Comment by jesserockz** [comment]
Maybe the yaml config should match too :thinking:

**Comment by jesserockz** [comment]
Im currently writing the HA core side too

**Comment by Cossid** [comment]
the "suggested" part didn't even occur to me until I saw the pre-existing constant, but I think I'd prefer to leave it as just `area` in the esphome config.  Aside from HA integration, I've got thoughts of using this field for organizing alternate dashboard listings, in which case "suggested" wouldn't really fit as well.

**Comment by jesserockz** [comment]
image

**Review by jesserockz — APPROVED** [review]

**Comment by nagyrobi** [comment]
It would have been nice to consider (perhaps an override) option for each component to be in a different area.

There are use cases when ESPHome acts as a central hub for switches, sensors, etc. these being spread out in different areas.

But I guess this needs support from HA side too..

**Comment by jesserockz** [comment]
@nagyrobi You can still move entities to a different area inside Home Assistant. They can be individually changed from the device area. 

This feature is only the "suggested area" that pre-fills the area selection when you first add a device to HA.

**Pull request merged**